services/galley/test/unit/Test/Galley/Mapping.hs

```diff
@@ -0,0 +1,209 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2020 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Test.Galley.Mapping where
+
+import Data.Domain
+import Data.Id
+import Data.Qualified
+import Galley.API ()
+import Galley.API.Mapping
+import qualified Galley.Data as Data
+-- import Galley.Types
+
+import Galley.Types (LocalMember, RemoteMember)
+import qualified Galley.Types.Conversations.Members as I
+import Imports
+import Test.Tasty
+import Test.Tasty.HUnit
+import Wire.API.Conversation
+import Wire.API.Conversation.Role (roleNameWireAdmin)
+
+tests :: TestTree
+tests =
+  testGroup
+    "ConversationMapping"
+    [ testCase "Alice@A Conv@A" runMappingSimple,
+      testCase "Alice@A Conv@A requester=not a member@A" runMappingNotAMemberA,
+      testCase "Alice@A Conv@A requester=not a member@B" runMappingNotAMemberB,
+      testCase "Alice@A Conv@A Bob@B" runMappingRemoteUser,
+      testCase "Alice@A Conv@B Bob@B" runMappingRemoteConv,
+      testCase "Alice@A Conv@B Bob@B bobUUID=aliceUUID" runMappingSameUnqualifiedUUID
+    ]
+
+runMappingSimple :: HasCallStack => IO ()
+runMappingSimple = do
+  let convDomain = Domain "backendA.example.com"
+  let userDomain = Domain "backendA.example.com"
+  alice <- randomId
+  let requester = Qualified alice userDomain
+  let expectedSelf = Just $ mkMember requester
+  let expectedOthers = Just []
+
+  let locals = [mkInternalMember requester]
+  let remotes = []
+  conv <- mkInternalConv locals remotes
+  let actual = cnvMembers <$> conversationViewMaybeQualified convDomain requester conv
+
+  assertEqual "self:" expectedSelf (cmSelf <$> actual)
+  assertEqual "others:" expectedOthers (cmOthers <$> actual)
+
+runMappingNotAMemberA :: HasCallStack => IO ()
+runMappingNotAMemberA = do
+  let convDomain = Domain "backendA.example.com"
+  let aliceDomain = Domain "backendA.example.com"
+  alice <- flip Qualified aliceDomain <$> randomId
+  requester <- flip Qualified aliceDomain <$> randomId
```

Comment:
An unstated assumption here is that `alice` and `requester` are two different (random) values. Sure, it is highly unlikely to happen, but there is a possibility of ending up with the same values here.

Given that randomness doesn't buy us anything in this case, and on the other hand can lead to trouble as I just pointed out, I'd simply use fixed, distinct and specific IDs for `alice` and `requester`.

Reply:
I disagree here, we use UUIDs for a reason - because we believe the chances for collision are so small that we accept that risk. We do this everywhere in production code, so it doesn't make sense to me to be more careful in unit test than in production code.
There is a specific test further down testing the UUID-collision case between alice and bob from two different backends: This is arguably more likely as it can be maliciously crafted this way by a backend that is not well-behaved.